javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -804,13 +885,31 @@ pragma[noinline]
 private predicate flowThroughProperty(
   DataFlow::Node pred, DataFlow::Node succ, DataFlow::Configuration cfg, PathSummary summary
 ) {
-  exists(string prop, DataFlow::Node base, PathSummary oldSummary, PathSummary newSummary |
-    reachableFromStoreBase(prop, pred, base, cfg, oldSummary) and
-    loadStep(base, succ, prop, cfg, newSummary) and
+  exists(string prop, DataFlow::Node storeBase, DataFlow::Node loadBase, PathSummary oldSummary, PathSummary newSummary |
+    reachableFromStoreBase(prop, pred, storeBase, cfg, oldSummary) and
+    (storeBase = loadBase or existsCopyProperty(storeBase, loadBase, prop, cfg)) and
+    loadStep(loadBase, succ, prop, cfg, newSummary) and
     summary = oldSummary.append(newSummary)
   )
 }
 
+/**
+ * Holds if the property `prop` is copied from `fromNode` to `toNode` using at least 1 step.
+ *
+ * The recursion of this predicate has been unfolded once compared to a naive implementation in order to avoid having no constraint on `prop`.
+ * Therefore a caller of this predicate should also test whether the `toNode` and `fromNode` are equal.
+ */
+private predicate existsCopyProperty(DataFlow::Node fromNode, DataFlow::Node toNode, string prop, DataFlow::Configuration cfg) {
```

Comment:
Instead of having this predicate, it should be possible to use `isAdditionalCopyPropertyStep` in `reachableFromStoreBase` as a forwarding step (i.e. a disjunct similar to the one using `flowStep`). I think everything will "just work" if we do it there.

`flowThroughProperty` is for collapsing store->load sequences, but copy-property is essentially a load->store sequence. Handling it in there feels like the wrong place, and is probably why we end up with a clunky dependency on `getALocalSource`.

Reply:
You were right, putting it in `reachableFromStoreBase` made everything "just work"!